File: Codility/9_maximum_slice_problem/max_double_slice.cpp
// https://codility.com/programmers/task/max_double_slice_sum/

int solution(vector<int> &A) {
    vector<int> maxEndingSums(A.size());
    int maxEnding = 0;
    for (int i = 1; i < A.size() ; ++i)
    {
        maxEnding = max(0, maxEnding + A[i]);
        maxEndingSums[i] = maxEnding;
    }
    
    vector<int> maxStartingSums(A.size());
    int maxStarting = 0;
    for (int i = A.size() - 2; i >= 0 ; --i)
    {
        maxStarting = max(0, maxStarting + A[i]);
        maxStartingSums[i] = maxStarting;
    }
    
    int maxDoubleSliceSum = 0;
    for (size_t i = 1; i < A.size() - 1; ++i)
    {
        const int maxEnding = max(0, maxEndingSums[i - 1]);
        const int maxStarting = max(0, maxStartingSums[i + 1]);
        maxDoubleSliceSum = max(maxDoubleSliceSum, maxEnding + maxStarting);
    }
    
    return maxDoubleSliceSum;
}

File: Codility/9_maximum_slice_problem/max_double_slice.py
# https://codility.com/programmers/task/max_double_slice_sum/

def solution(A):
    max_ending = [0] * len(A)
    for i in xrange(1, len(A)):
        max_ending[i] = max(0, max_ending[i - 1] + A[i])

    max_starting = [0] * len(A)
    for i in xrange(len(A) - 2, -1, -1):
        max_starting[i] = max(0, max_starting[i + 1] + A[i])
        
    max_double_slice = 0
    for i in xrange(1, len(A) - 1):
        max_double_slice = max(max_double_slice, max_ending[i - 1] + max_starting[i + 1])
        
    return max_double_slice
